### PR TITLE
[PLATFORM-865]: Add a builder type for redaction to Veil, for manually redacting arbitrary data

### PR DESCRIPTION
**Pull request by WilliamVenner**
https://prima-assicurazioni-spa.myjetbrains.com/youtrack/issue/PLATFORM-865

This PR introduces the `Redactor`, `RedactorBuilder` and `RedactWrapped` types.

The `RedactorBuilder` allows for building a `Redactor` using familiar flags seen in the `#[redact]` derive macro. The built `Redactor` may then be used to redact any arbitrary data using the flags provided.

The `RedactorBuilder` has been set up for 100% const evaluation, therefore the compiler is able to completely eliminate any error checking at compile time. The error checking is zero cost if everything is known at compile time :D

Finally I have also used the `RedactWrapped` type as a convenience API. It is returned by the `wrap` method of `Redactor`. I have documented this with examples, and plan to document everything on a higher level (e.g. in examples/ and the crate root docs) in a separate PR.

### CODE REVIEW

File: src/builder.rs
Comment (MaeIsBad) [inline src/builder.rs:1]: The name builder for the module feels a bit too generic here, maybe we can call it something like "redaction_builder" instead?

File: src/builder.rs
Comment (MaeIsBad) [inline src/builder.rs:58]: I would suggest adding a "builder" associated function here that returns a RedactorBuilder

File: src/lib.rs
Comment (MaeIsBad) [inline src/lib.rs:243]: should all of those be exposed on the top level? I feel like RedactorBuilder and RedactWrapped aren't types that need to be named often, and could be shoved into a submodule instead

File: src/builder.rs
Comment (WilliamVenner) [reply src/builder.rs:169]: No no, I checked this one with Godbolt. It doesn't optimise away the compile time checks without the inlines.

Comment (MaeIsBad) [reply src/builder.rs:169]: that doesn't seem to be the case even with opt_level 1.
https://godbolt.org/z/5WqMaqvav

I don't think we care about inlining here in dev, since it might slow down compilation

Comment (WilliamVenner) [reply src/builder.rs:169]: https://godbolt.org/z/nfxcK7EKs

If you take the `#[inline(always)]` away from `build`, you can see that it just fails to const evaluate for some reason

Comment (MaeIsBad) [reply src/builder.rs:169]: I have an absurd level of faith in compilers, so my heart is telling me we don't need to do this, but I can't argue with the facts.

Comment (WilliamVenner) [reply src/builder.rs:169]: This particular case also caused me psychological damage at the time

I'm here for you if you need support

File: src/builder.rs
Comment (WilliamVenner) [reply src/builder.rs:52]: oh yeah, I assumed the derive would add `where T: Clone` and stuff, but it doesn't

File: src/builder.rs
Comment (WilliamVenner) [reply src/builder.rs:1]: Decided to go for just `redactor` instead :)

File: src/redactor.rs
Comment (MaeIsBad) [inline src/redactor.rs:69]: Would it be better to use `Redactor::builder` instead of `RedactorBuilder::new` here?

Comment (WilliamVenner) [reply src/redactor.rs:69]: Sure, that would be nicer

**Review by MaeIsBad — APPROVED** [review]
I think this is good to merge, but I'd wait for another review on all these PRs

**Review by MaeIsBad — APPROVED** [review]

**Pull request merged**
